test/cctest/test_inspector_socket.cc

```diff
@@ -0,0 +1,849 @@
+    uv_run(&loop, UV_RUN_NOWAIT);
+  }
+
+  uv_close((uv_handle_t *)&client_socket, nullptr);
+  bool flag = false;
+  inspector.data = &flag;
+  SPIN_WHILE(!flag);
+  inspector.data = nullptr;
+}
+
+TEST_F(InspectorSocketTest, EOFBeforeHandshake) {
+  const char MESSAGE[] = "We'll send EOF afterwards";
+  inspector_write(&inspector, MESSAGE, sizeof(MESSAGE) - 1);
+  expect_on_client(MESSAGE, sizeof(MESSAGE) - 1);
+  uv_close((uv_handle_t *)&client_socket, nullptr);
+  SPIN_WHILE(last_event != kInspectorHandshakeFailed);
+}
+
+static void fill_message(char* buffer, size_t len) {
+  buffer[len - 1] = '\0';
+  for (size_t i = 0; i < len - 1; i++) {
+    buffer[i] = 'a' + (i % ('z' - 'a'));
+  }
+}
+
+static void mask_message(const char *message,
+                         char* buffer, const char mask[]) {
+  const size_t mask_len = 4;
+  int i = 0;
+  while (*message != '\0') {
+    *buffer++ = *message++ ^ mask[i++ % mask_len];
+  }
+}
+
+TEST_F(InspectorSocketTest, Send1Mb) {
+  ASSERT_TRUE(connected);
+  ASSERT_FALSE(inspector_ready);
+  do_write((char *)HANDSHAKE_REQ, sizeof(HANDSHAKE_REQ) - 1);
+  SPIN_WHILE(!inspector_ready);
+  expect_handshake();
+
+  const size_t message_len = 1000000;
+
+  // 2. Brief exchange
+  char* message = (char*) malloc(message_len + 1);
+  fill_message(message, message_len + 1);
+
+  // 1000000 is 0xF4240 hex
+  const char EXPECTED_FRAME_HEADER[] = {
+    '\x81', '\x7f', '\x00', '\x00', '\x00', '\x00', '\x00', '\x0F',
+    '\x42', '\x40'
+  };
+  char* expected = (char*) malloc(sizeof(EXPECTED_FRAME_HEADER) + message_len);
+
+  memcpy(expected, EXPECTED_FRAME_HEADER, sizeof(EXPECTED_FRAME_HEADER));
+  memcpy(expected + sizeof(EXPECTED_FRAME_HEADER), message, message_len);
+
+  inspector_write(&inspector, message, message_len);
+  expect_on_client(expected, sizeof(EXPECTED_FRAME_HEADER) + message_len);
+
+  char MASK[4] = {'W', 'h', 'O', 'a'};
+
+  const char FRAME_TO_SERVER_HEADER[] = {
+    '\x81', '\xff', '\x00', '\x00', '\x00', '\x00', '\x00', '\x0F',
+    '\x42', '\x40', MASK[0], MASK[1], MASK[2], MASK[3]
+  };
+
+  const size_t outgoing_len = sizeof(FRAME_TO_SERVER_HEADER) + message_len;
+  char *outgoing = (char*) malloc(outgoing_len);
+  memcpy(outgoing, FRAME_TO_SERVER_HEADER, sizeof(FRAME_TO_SERVER_HEADER));
+  mask_message(message, outgoing + sizeof(FRAME_TO_SERVER_HEADER), MASK);
+
+  setup_inspector_expecting(); // Buffer on the client side.
+  do_write(outgoing, outgoing_len);
+  expect_on_server(message, message_len);
+
+  // 3. Close
+  const char CLIENT_CLOSE_FRAME[] = {'\x88', '\x80', '\x2D',
+                                     '\x0E', '\x1E', '\xFA'};
+  const char SERVER_CLOSE_FRAME[] = {'\x88', '\x00'};
+  do_write(CLIENT_CLOSE_FRAME, sizeof(CLIENT_CLOSE_FRAME));
+  expect_on_client(SERVER_CLOSE_FRAME, sizeof(SERVER_CLOSE_FRAME));
+  GTEST_ASSERT_EQ(0, uv_is_active((uv_handle_t *)&client_socket));
+  free(outgoing);
+  free(expected);
+  free(message);
+}
```

Comment:
Can you `uv_loop_close(&loop)` on exit?  Valgrind complains about memory leaks now.